src/awkward/_v2/contents/indexedarray.py

```diff
@@ -280,6 +280,105 @@ def _combinations(self, n, replacement, recordlookup, parameters, axis, depth):
                 n, replacement, recordlookup, parameters, posaxis, depth
             )
 
+    def _reduce_next(
+        self,
+        reducer,
+        negaxis,
+        starts,
+        shifts,
+        parents,
+        outlength,
+        mask,
+        keepdims,
+    ):
+        nplike = self.nplike
+        branch, depth = self.branch_depth
+
+        index_length = len(self._index)
+
+        nextcarry = ak._v2.index.Index64.zeros(index_length, nplike)
+        nextparents = ak._v2.index.Index64.zeros(index_length, nplike)
+        outindex = ak._v2.index.Index64.zeros(index_length, nplike)
+        self._handle_error(
+            nplike[
+                "awkward_IndexedArray_reduce_next_64",
+                nextcarry.dtype.type,
+                nextparents.dtype.type,
+                outindex.dtype.type,
+                self._index.dtype.type,
+                parents.dtype.type,
+            ](
+                nextcarry.to(nplike),
+                nextparents.to(nplike),
+                outindex.to(nplike),
+                self._index.to(nplike),
+                parents.to(nplike),
+                index_length,
+            )
+        )
+        next = self._content._carry(nextcarry, False, NestedIndexError)
+        nextshifts = None
+        out = next._reduce_next(
+            reducer,
+            negaxis,
+            starts,
+            nextshifts,
+            nextparents,
+            outlength,
+            mask,
+            keepdims,
+        )
+
+        if not branch and negaxis == depth:
+            return out
+        else:
+            if isinstance(out, ak._v2.contents.RegularArray):
+                out = out.toListOffsetArray64(True)
+
+            elif isinstance(out, ak._v2.contents.ListOffsetArray):
+                if len(starts) > 0 and starts[0] != 0:
+                    raise AssertionError(
+                        "reduce_next with unbranching depth > negaxis expects a "
+                        "ListOffsetArray64 whose offsets start at zero ({0})".format(
+                            starts[0]
+                        )
+                    )
+
+                outoffsets = ak._v2.index.Index64.empty(len(starts) + 1, nplike)
+                self._handle_error(
+                    nplike[
+                        "awkward_IndexedArray_reduce_next_fix_offsets_64",
+                        outoffsets.dtype.type,
+                        starts.dtype.type,
+                    ](
+                        outoffsets.to(nplike),
+                        starts.to(nplike),
+                        len(starts),
+                        len(self._index),
+                    )
+                )
+
+                tmp = ak._v2.contents.IndexedOptionArray(
+                    outindex,
+                    out._content,
+                    None,
+                    None,
+                )._simplify_optiontype()
```

Comment:
For IndexedArray (this file) and _not_ for IndexedOptionArray (which are separate files in v2), this `tmp` should be created as an IndexedArray.

That was a bug-fix in v1 that, unfortunately, I didn't fix before you copied this code into v2. It was fixed in PR #1108 ([see diff](https://github.com/scikit-hep/awkward-1.0/pull/1108/files)), which has a test you can directly adapt for v2:

https://github.com/scikit-hep/awkward-1.0/blob/fe103450d6001d6fb84382c0af79c4dde501a26c/tests/test_1071-mask-identity-false-should-not-return-option-type.py#L11-L22

Leaving the code above unchanged, the assertion would fail (the generated type would have a "?" in it, but the comparison does not), but changing the above IndexedOptionArray into an IndexedArray would make the assertion pass.

In v2, you can access the `type` through `.form.type` because the high-level `ak.Array` isn't set up for it yet.

Reply:
done